src/tools/ScenarioMeasurement/Startup/TimeToMain2Parser.cs

```diff
@@ -0,0 +1,108 @@
+
+    public void EnableKernelProvider(ITraceSession kernel)
+    {
+        kernel.EnableKernelProvider(TraceSessionManager.KernelKeyword.Process, TraceSessionManager.KernelKeyword.Thread, TraceSessionManager.KernelKeyword.ContextSwitch);
+    }
+
+    public void EnableUserProviders(ITraceSession user)
+    {
+        user.EnableUserProvider(EventSourceName, TraceEventLevel.Verbose);
+    }
+
+    public IEnumerable<Counter> Parse(string mergeTraceFile, string processName, IList<int> pids, string commandLine)
+    {
+        var results = new List<double>();
+        var threadTimes = new List<double>();
+        double threadTime = 0;
+        var ins = new Dictionary<int, double>();
+        double start = -1;
+        int? pid = null;
+        using (var source = new TraceSourceManager(mergeTraceFile))
+        {
+            source.Kernel.ProcessStart += evt =>
+            {
+                if (!pid.HasValue && ParserUtility.MatchProcessStart(evt, source, processName, pids, commandLine))
+                {
+                    pid = evt.ProcessID;
+                    start = evt.TimeStampRelativeMSec;
+                }
+            };
+
+            source.Source.Kernel.ThreadCSwitch += evt =>
+            {
+                if (!pid.HasValue) // we're currently in a measurement interval
+                    return;
+
+                if (evt.NewProcessID != pid && evt.OldProcessID != pid)
+                    return; // but this isn't our process
+
+                if (evt.OldProcessID == pid) // this is a switch out from our process
+                {
+                    if (ins.TryGetValue(evt.OldThreadID, out var value)) // had we ever recorded a switch in for this thread?
+                    {
+                        threadTime += evt.TimeStampRelativeMSec - value;
+                        ins.Remove(evt.OldThreadID);
+                    }
+                }
+                else // this is a switch in to our process
+                {
+                    ins[evt.NewThreadID] = evt.TimeStampRelativeMSec;
+                }
+            };
+
+            if (source.IsWindows)
+            {
+                source.Source.Dynamic.AddCallbackForProviderEvent(EventSourceName, OnMainEventName, evt =>
+                {
+                    if (pid.HasValue && evt.ProcessID == pid && evt.ProcessName.Equals(processName, StringComparison.OrdinalIgnoreCase))
+                    {
+                        results.Add(evt.TimeStampRelativeMSec - start);
+                        threadTimes.Add(threadTime);
+                        pid = null;
+                        threadTime = 0;
+                        start = 0;
+                    }
+                });
+            }
+            else
+            {
+                source.Source.Clr.EventSourceEvent += evt =>
+                {
+                    if (pid.HasValue && evt.ProcessID == pid && evt.EventID == OnMainEventId)
```

Comment:
Just for my own understanding. This `EventSourceEvent` is only activated if you register your own events with `EventSource`, and then enable them and they fire. In that way we know that only our events will show up here and there will not be any overlap of `EventID` with another event. Is that the correct understanding? If so should we add a comment that explains that all of our custom events need to have unique IDs across all events?

Reply:
We're setting the `COMPlus_EventSourceFilter`.

Reply:
OK, so we can only ever use one at a time anyone. I understand now.

Reply:
I've added also `MagicConstant`, to "prevent" collisions.